crates/node-core/src/init.rs

```diff
@@ -233,6 +281,123 @@ pub fn insert_genesis_header<DB: Database>(
     Ok(())
 }
 
+/// Initialize chain with state at specific block, from reader of state dump.
+pub fn init_from_state_dump<DB: Database>(
+    mut reader: impl BufRead,
+    factory: ProviderFactory<DB>,
+) -> eyre::Result<B256> {
+    let block = factory.last_block_number()?;
+    let hash = factory.block_hash(block)?.unwrap();
+
+    debug!(target: "reth::cli",
+        block,
+        chain=%factory.chain_spec().chain,
+        "Initializing state at block"
+    );
+
+    let mut total_inserted_accounts = 0;
+    let mut accounts = Vec::with_capacity(AVERAGE_COUNT_ACCOUNTS_PER_GB_STATE_DUMP);
+    let mut chunk_total_byte_len = 0;
+    let mut line = String::new();
+
+    // first line can be state root, then it can be used for verifying against computed state root
+    reader.read_line(&mut line)?;
+    let _state_root = match serde_json::from_str::<StateRoot>(&line) {
+        Ok(root) => {
+            trace!(target: "reth::cli",
+                root=%root.root,
+                "Read state root from file"
+            );
+            Some(root)
+        }
+        Err(_) => {
+            let GenesisAccountWithAddress { genesis_account, address } =
+                serde_json::from_str(&line)?;
+            accounts.push((address, genesis_account));
+
+            None
+        }
+    };
+    line.clear();
+
+    // remaining lines are accounts
+    while let Ok(n) = reader.read_line(&mut line) {
+        chunk_total_byte_len += n;
+        if DEFAULT_SOFT_LIMIT_BYTE_LEN_ACCOUNTS_CHUNK <= chunk_total_byte_len || n == 0 {
+            // acc
+            total_inserted_accounts += accounts.len();
+
+            debug!(target: "reth::cli",
+                chunk_total_byte_len,
+                parsed_new_accounts=accounts.len(),
+                total_inserted_accounts,
+                "Writing accounts to db"
+            );
+
+            // reset
+            chunk_total_byte_len = 0;
+
+            // use transaction to insert genesis header
+            let provider_rw = factory.provider_rw()?;
+            insert_genesis_hashes(
+                &provider_rw,
+                accounts.iter().map(|(address, account)| (address, account)),
+            )?;
+            insert_history(
+                &provider_rw,
+                accounts.iter().map(|(address, account)| (address, account)),
+                block,
+            )?;
+
+            // block is already written to static files
+            let tx = provider_rw.into_tx();
+
+            insert_state::<DB>(
+                &tx,
+                accounts.len(),
+                accounts.iter().map(|(address, account)| (address, account)),
+                block,
+                false,
+            )?;
+
+            // insert sync stage
+            for stage in StageId::ALL.iter() {
+                tx.put::<tables::StageCheckpoints>(stage.to_string(), StageCheckpoint::new(block))?;
+            }
+
+            tx.commit()?;
```

Comment:
doing a commit here will lead to an inconsistent state if the command crashes or is killed before the import is successful